### PR TITLE
Fixtures

### PR DESCRIPTION
**Pull request by dantleech**
This PR introduces a fixture loader similar to the DoctrineFixtures loader.

Load fixtures from all bundles:

````bash
$ php app/console sulu:document:fixtures:load
````

By default it will purge the workspace and initialize the workspace prior to loading the fixtures.

Fixtures take the same form as the DoctrineFixtures bundle:

````php
<?php

class FooFixtures implements DocumentFixtureInterface
{
    public function getOrder()
    {
        return 10;
    }

    public function load(DocumentManager $documentManager)
    {
        $document = $documentMananger->create('page');
        $document->setTitle('Hello');
        $document->getContent()->bind(array(
            'body' => 'Hello',
            'medias' => array(12, 12, 12),
        ));

        $documentManager->persist($document, 'de', array(
            'path' => '/foo/bar'
        ));
        $documentManager->flush();
    }
}
````

Added documentation https://github.com/sulu-io/sulu-docs/blob/document-manager/reference/components/document-manager/fixtures.rst

### CODE REVIEW

File: src/Sulu/Bundle/ContentBundle/DataFixtures/Document/FooFixture.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<?php
+
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/DataFixtures/Document/FooFixture.php]: Remove this file.

File: src/Sulu/Bundle/DocumentManagerBundle/Command/FixturesLoadCommand.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+use Symfony\Component\HttpKernel\KernelInterface;
+
+class FixturesLoadCommand extends Command
+{
+    private $loader;
+    private $executor;
+    private $kernel;
+
+
+    public function __construct(
```
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/Command/FixturesLoadCommand.php]: Add docs and remove empty lines.

File: src/Sulu/Bundle/DocumentManagerBundle/DataFixtures/DocumentExecutor.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+namespace Sulu\Bundle\DocumentManagerBundle\DataFixtures;
+
+use Symfony\Component\Console\Output\NullOutput;
+use Sulu\Bundle\DocumentManagerBundle\Initializer\Initializer;
+use Sulu\Component\DocumentManager\NodeManager;
+use Sulu\Component\DocumentManager\DocumentManager;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class DocumentExecutor
```
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/DataFixtures/DocumentExecutor.php]: Add docs

File: src/Sulu/Bundle/DocumentManagerBundle/DataFixtures/DocumentFixtureLoader.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
+<?php
+
+namespace Sulu\Bundle\DocumentManagerBundle\DataFixtures;
+
+use Sulu\Bundle\DocumentManagerBundle\DataFixtures\DocumentFixtureInterface;
+use Symfony\Component\DependencyInjection\ContainerAwareInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\Finder\Finder;
+
+class DocumentFixtureLoader
```
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/DataFixtures/DocumentFixtureLoader.php]: Add docs.

File: src/Sulu/Bundle/DocumentManagerBundle/DataFixtures/DocumentFixtureLoader.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    public function __construct(ContainerInterface $container)
+    {
+        $this->container = $container;
+    }
+
+    public function load(array $paths)
+    {
+        $finder = new Finder();
+        $finder->in($paths);
+        $finder->name('*Fixture.php');
```
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/DataFixtures/DocumentFixtureLoader.php]: Document this.

File: src/Sulu/Bundle/DocumentManagerBundle/Resources/config/command.xml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+            <argument type="service" id="sulu_document_manager.data_fixtures.executor" />
+            <argument type="service" id="kernel" />
+            <tag name="console.command" />
+        </service>
+
+    </services>
+</container>
+
+
+
```
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/Resources/config/command.xml]: Remove extra lines.

File: src/Sulu/Bundle/DocumentManagerBundle/Resources/config/data_fixtures.xml [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+            <argument type="service" id="sulu_document_manager.node_manager" />
+            <argument type="service" id="sulu_document_manager.initializer" />
+        </service>
+
+    </services>
+</container>
+
+
+
+
```
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/Resources/config/data_fixtures.xml]: remove extra lines.

File: src/Sulu/Bundle/DocumentManagerBundle/Tests/Unit/Command/FixturesLoadCommandTest.php [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+use Symfony\Component\HttpKernel\Bundle\BundleInterface;
+use Sulu\Bundle\DocumentManagerBundle\DataFixtures\DocumentFixtureInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Prophecy\Argument;
+
+class FixturesLoadCommandTest extends \PHPUnit_Framework_TestCase
+{
+    public function setUp()
+    {
+
```
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/Tests/Unit/Command/FixturesLoadCommandTest.php]: Extra/ Line/

File: src/Sulu/Bundle/DocumentManagerBundle/Tests/Unit/Command/FixturesLoadCommandTest.php [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+        $command = $application->find('sulu:document:fixtures:load');
+        $this->commandTester = new CommandTester($command);
+
+        $this->kernel->getBundles()->willReturn(array(
+            $this->bundle->reveal()
+        ));
+        $this->bundle->getPath()->willReturn(
+            __DIR__ . '/bundle'
+        );
+
```
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/Tests/Unit/Command/FixturesLoadCommandTest.php]: sgsin

File: src/Sulu/Bundle/DocumentManagerBundle/Tests/Unit/DataFixtures/DocumentFixtureLoaderTest.php [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    private $container;
+
+    public function setUp()
+    {
+        $this->container = $this->prophesize(ContainerInterface::class);
+        $this->loader = new DocumentFixtureLoader($this->container->reveal());
+    }
+
+    /**
+     * It should load, instantiate and order fixture classes
```
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/Tests/Unit/DataFixtures/DocumentFixtureLoaderTest.php]: It should also set the container when it implements ContainerAwareInterface

File: src/Sulu/Bundle/DocumentManagerBundle/Tests/Unit/DataFixtures/fixtures/ContainerFixture.php [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    public function load(DocumentManager $documentManager)
+    {
+    }
+
+    public function getOrder()
+    {
+        return 30;
+    }
+}
+
```
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/Tests/Unit/DataFixtures/fixtures/ContainerFixture.php]: Extra

File: src/Sulu/Bundle/DocumentManagerBundle/Tests/foo.yml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+alias: page
+structure_type: overview
+content:
+    title: Foobar
+    article: bar
```
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/Tests/foo.yml]: Remove.

**Comment by dantleech** [comment]
Updated.

File: src/Sulu/Bundle/DocumentManagerBundle/Command/FixturesLoadCommand.php [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+     * @var KernelInterface
+     */
+    private $kernel;
+
+    public function __construct(
+        DocumentFixtureLoader $loader,
+        DocumentExecutor $executor,
+        KernelInterface $kernel
+    )
+    {
```
Comment (danrot) [inline src/Sulu/Bundle/DocumentManagerBundle/Command/FixturesLoadCommand.php]: CS: Previous two lines should go into one

File: src/Sulu/Bundle/DocumentManagerBundle/Command/FixturesLoadCommand.php [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+     * {@inheritDoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $noInitialize = $input->getOption('no-initialize');
+        $append = $input->getOption('append');
+
+        if ($input->isInteractive() && !$append) {
+            $dialog = $this->getHelperSet()->get('dialog');
+            if (!$dialog->askConfirmation($output, '<question>Careful, database will be purged. Do you want to continue Y/N ?</question>', false)) {
```
Comment (danrot) [inline src/Sulu/Bundle/DocumentManagerBundle/Command/FixturesLoadCommand.php]: line too long?

File: src/Sulu/Bundle/DocumentManagerBundle/DataFixtures/DocumentFixtureInterface.php [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+     *
+     * Use the document manager to create and save fixtures.
+     * Be sure to call DocumentManager#save() when you are done.
+     *
+     * @param DocumentManager
+     */
+    public function load(DocumentManager $documentManager);
+
+    /**
+     * Return the order relative to other fixture orders.
```
Comment (danrot) [inline src/Sulu/Bundle/DocumentManagerBundle/DataFixtures/DocumentFixtureInterface.php]: This order is absolute, isn't it? IMO relative might be confusing here...

File: src/Sulu/Bundle/DocumentManagerBundle/DataFixtures/DocumentExecutor.php (hunk 3 above)
Comment (danrot) [inline src/Sulu/Bundle/DocumentManagerBundle/DataFixtures/DocumentExecutor.php]: Did you overlook this comment? :-)

File: src/Sulu/Bundle/DocumentManagerBundle/DataFixtures/DocumentFixtureLoader.php [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+        foreach ($finder as $file) {
+            $declaredClasses = get_declared_classes();
+            require_once($file);
+            $declaredClassesDiff = array_diff(get_declared_classes(), $declaredClasses);
+            $fixtureClass = array_pop($declaredClassesDiff);
```
Comment (danrot) [inline src/Sulu/Bundle/DocumentManagerBundle/DataFixtures/DocumentFixtureLoader.php]: Just for me to understand: There's another restriction, in the fixture file may only be one file, right? (I know that this also covered by PSR)

Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/DataFixtures/DocumentFixtureLoader.php]: one class you mean? yeah. tbh determining the class from a file is non-trivial. The above method will not work twice in the same request for instance. The doctrine fixtures bundle just traverses ALL of the loaded classes and instanatiates ALL which implement the DoctrineFixturesInterface. But that approach caused issues here with prophecy.

I think this method will be ok.

File: src/Sulu/Bundle/DocumentManagerBundle/Tests/Unit/Command/bundle/DataFixtures/Document/FooFixture.php [hunk 17]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+namespace Sulu\Bundle\DocumentManagerBundle\Tests\Unit\Command\bundle\DataFixtures\Document;
```
Comment (danrot) [inline src/Sulu/Bundle/DocumentManagerBundle/Tests/Unit/Command/bundle/DataFixtures/Document/FooFixture.php]: `Bundle` instead of `bundle`

Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/Tests/Unit/Command/bundle/DataFixtures/Document/FooFixture.php]: Its a fixture

Comment (danrot) [inline src/Sulu/Bundle/DocumentManagerBundle/Tests/Unit/Command/bundle/DataFixtures/Document/FooFixture.php]: Ok, then let's call this directory `fixtures`

File: src/Sulu/Bundle/DocumentManagerBundle/Command/FixturesLoadCommand.php (hunk 14 above)
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/Command/FixturesLoadCommand.php]: Yeah, this was adapted code from the PHPCRBundle.

**Comment by dantleech** [comment]
Fixed the outstanding doc issue.

**Pull request merged**